x/stake/keeper/validator.go

```diff
@@ -69,34 +69,33 @@ func (k Keeper) GetAllValidators(ctx sdk.Context) (validators []types.Validator)
 	iterator := sdk.KVStorePrefixIterator(store, ValidatorsKey)
 
 	i := 0
-	for ; ; i++ {
-		if !iterator.Valid() {
-			break
-		}
+	for ; iterator.Valid(); iterator.Next() {
 		addr := iterator.Key()[1:]
 		validator := types.MustUnmarshalValidator(k.cdc, addr, iterator.Value())
 		validators = append(validators, validator)
-		iterator.Next()
+		i++
 	}
 	iterator.Close()
 	return validators
 }
 
 // Get the set of all validators, retrieve a maxRetrieve number of records
-func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve int16) (validators []types.Validator) {
+func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve ...int16) (validators []types.Validator) {
+	if len(maxRetrieve) == 0 {
+		return k.GetAllValidators(ctx)
+	}
+	validators = make([]types.Validator, maxRetrieve[0])
+
 	store := ctx.KVStore(k.storeKey)
 	iterator := sdk.KVStorePrefixIterator(store, ValidatorsKey)
 
-	validators = make([]types.Validator, maxRetrieve)
 	i := 0
-	for ; ; i++ {
-		if !iterator.Valid() || i > int(maxRetrieve-1) {
-			break
-		}
+	for ; iterator.Valid() && i < int(maxRetrieve[0]); iterator.Next() {
```

Comment:
`for i := 0; iterator.Valid() && i < int(maxRetrieve[0]); iterator.Next() {`

Reply:
not possible. `return validators[:i] // trim` throws `i` is undefined err